pkg/build/registry/rest.go

```diff
@@ -22,39 +21,38 @@ var (
 // WaitForRunningBuild waits until the specified build is no longer New or Pending. Returns true if
 // the build ran within timeout, false if it did not, and an error if any other error state occurred.
 // The last observed Build state is returned.
-func WaitForRunningBuild(watcher rest.Watcher, ctx apirequest.Context, build *buildapi.Build, timeout time.Duration) (*buildapi.Build, bool, error) {
+func WaitForRunningBuild(buildClient buildtypedclient.BuildsGetter, build *buildapi.Build, timeout time.Duration) (*buildapi.Build, bool, error) {
 	fieldSelector := fields.OneTermEqualSelector("metadata.name", build.Name)
-	options := &metainternal.ListOptions{FieldSelector: fieldSelector, ResourceVersion: build.ResourceVersion}
-	w, err := watcher.Watch(ctx, options)
+	options := metav1.ListOptions{FieldSelector: fieldSelector.String(), ResourceVersion: build.ResourceVersion}
+	w, err := buildClient.Builds(build.Namespace).Watch(options)
 	if err != nil {
 		return build, false, err
 	}
-	defer w.Stop()
 
 	observed := build
-	ch := w.ResultChan()
-	expire := time.After(timeout)
-	for {
-		select {
-		case event := <-ch:
-			obj, ok := event.Object.(*buildapi.Build)
-			if !ok {
-				return observed, false, fmt.Errorf("received unknown object while watching for builds")
-			}
-			observed = obj
-
-			if event.Type == watch.Deleted {
-				return observed, false, ErrBuildDeleted
-			}
-			switch obj.Status.Phase {
-			case buildapi.BuildPhaseRunning, buildapi.BuildPhaseComplete, buildapi.BuildPhaseFailed, buildapi.BuildPhaseError, buildapi.BuildPhaseCancelled:
-				return observed, true, nil
-			case buildapi.BuildPhaseNew, buildapi.BuildPhasePending:
-			default:
-				return observed, false, ErrUnknownBuildPhase
-			}
-		case <-expire:
-			return observed, false, nil
+	_, err = watch.Until(timeout, w, func(event watch.Event) (bool, error) {
+		obj, ok := event.Object.(*buildapi.Build)
+		if !ok {
+			return false, fmt.Errorf("received unknown object while watching for builds: %T", event.Object)
 		}
+		observed = obj
+
+		if event.Type == watch.Deleted {
+			return false, ErrBuildDeleted
+		}
+		switch obj.Status.Phase {
+		case buildapi.BuildPhaseRunning, buildapi.BuildPhaseComplete, buildapi.BuildPhaseFailed, buildapi.BuildPhaseError, buildapi.BuildPhaseCancelled:
+			return true, nil
+		case buildapi.BuildPhaseNew, buildapi.BuildPhasePending:
+		default:
+			return false, ErrUnknownBuildPhase
+		}
+
+		return false, nil
+	})
+	if err != nil {
+		return nil, false, err
 	}
+
+	return observed, true, nil
```

Comment:
@deads2k @gabemontero if i'm reading this correctly, we now return "observed, true, nil" if we hit the timeout, whereas previously we returned "observed, false, nil".

and presumably if we hit a timeout, "observed" will be nil.

Reply:
(hm, maybe not, seems like watch.Until is supposed to be returning an error if it times out)

Reply:
ah, but we used to always return the input build if we timed out, now we return nil.